### ISSUE TITLE
Image overlay inside cells

### ISSUE THREAD

**Issue by knorrest** [clj-pdf/clj-pdf#165]
I'm creating pdf:s with the library, using borderless nested pdf-tables to make it simple to place my content. My problem is that in addition to text and background, I would need to add image overlays inside some cells. In openpdf/itext such things seem to be done using cell events. Am I right in that there is no support for this in the library?

**Comment by yogthos** [comment]
Yeah, there's no support for hooking into cell events at the moment. I'd be open to a PR for declaring an events hook in the table cell options map if you'd be interested to take a look.

**Comment by knorrest** [comment]
Ok, I'll look at it - will get back to you when I've thought a bit...

**Comment by knorrest** [comment]
Ok, made a PR!

**Comment by knorrest** [comment]
One option for putting images in cells would be to add some code I now have in my app to the library:

`
(defn- h-align [alignment left width img-left]
  (case alignment
    :left left
    :right (+ left (- width img-left))
    (+ left (/ (- width img-left) 2))))

(defn- v-align [alignment bottom height img-height]
  (case alignment
    :bottom bottom
    :top (+ bottom (- height img-height))
    (+ bottom (/ (- height img-height) 2))))

(defn- fit-image-fn [halignment valignment ^URL img-url]
  (fn [^Rectangle pos ^PdfContentByte canvas]
    (let [left (.getLeft pos)
          bottom (.getBottom pos)
          width (.getWidth pos)
          height (.getHeight pos)
          img (Image/getInstance img-url)]
      (.scaleToFit img width height)
      (let [img-width (.getScaledWidth img)
            img-height (.getScaledHeight img)
            img-left (h-align halignment left width img-width)
            img-bottom (v-align valignment bottom height img-height)]
        (.setAbsolutePosition img img-left img-bottom)
        (.addImage canvas img)))))`

Used like:

`{:background-layer-fn (fit-image-fn :center :bottom trend-image)}`

**Comment by yogthos** [comment]
Thanks, just saw the PR, and trying it out locally everything looks good to me. I think the helper functions would be useful to add. Might make sense to expose them declaratively via cell options. Maybe something like `{:background-layer {:fit-image? true ...}`.

**Comment by knorrest** [comment]
Ok, I’ll sketch something up and get back.

**Comment by yogthos** [comment]
Fantastic, and I'm good to do a release as soon as we get this ironed out.

**Comment by knorrest** [comment]
Thinking about it, maybe the best option is to just add a section in the readme with some examples. Theres a million things that one could do in cell event callbacks, with image background or overlay and shading being the ones that I needed now. A few simple examples might be enough to just show the idea. I agree that the declarative approach is really nice, but since we're using the more "advanced" part of the underlying API here it might make sense to just expose that in a nice way.

**Comment by yogthos** [comment]
I'm ok with that, if some common patterns emerge we can add declarative syntax to cover those via the standard API at a later point. I can push out a release tonight then.

**Comment by knorrest** [comment]
Ok, I'll add something to the doc and make a  PR. If you want that in the release you'll have to wait until tomorrow, I'm a bit stuck with other things today.

**Comment by knorrest** [comment]
Found some time, and did a PR with example added to the readme. :-)

**Comment by knorrest** [comment]
...and a release would be terrific!

**Comment by yogthos** [comment]
Fantastic, thanks for adding the docs and I just pushed out `2.3.2` with the updates.
